Concurrent AWS testing collides

The way we run AWS tests right now means that there is a chance concurrent tests can collide.  For example, we seem to have hit this in https://travis-ci.com/pulumi/lumi/builds/48847447:

```
[Provider Test]: Getting aws:dynamodb/table:Table with id arn:aws:dynamodb:us-west-2:153052954103:table/lumitest-430c81b7d7a1c7d16cb5096529ac2a7f79e2c814
[Provider Test]: Checking aws:dynamodb/table:Table
[Provider Test]: Updating aws:dynamodb/table:Table with id arn:aws:dynamodb:us-west-2:153052954103:table/lumitest-430c81b7d7a1c7d16cb5096529ac2a7f79e2c814
Updating provisioned capacity for DynamoDB Table 0xc420439f70
Adding new global secondary index myGSI2 for DynamoDB Table lumitest-430c81b7d7a1c7d16cb5096529ac2a7f79e2c814
[Provider Test]: Deleting aws:dynamodb/table:Table with id arn:aws:dynamodb:us-west-2:153052954103:table/lumitest-430c81b7d7a1c7d16cb5096529ac2a7f79e2c814
Deleting DynamoDB Table 'lumitest-430c81b7d7a1c7d16cb5096529ac2a7f79e2c814'
DynamoDB Table delete request submitted; waiting for it to delete
--- FAIL: Test (161.54s)
	Error Trace:	table_test.go:24
	Error:      	Expected nil, but got: &awserr.requestError{awsError:(*awserr.baseError)(0xc4203441c0), statusCode:400, requestID:"EEFOJ0AFV03HOQ3ACDMIR5A5VBVV4KQNSO5AEMVJF66Q9ASUAAJG"}
FAIL
coverage: 69.4% of statements
FAIL	github.com/pulumi/lumi/lib/aws/provider/dynamodb	161.552s
ok  	github.com/pulumi/lumi/lib/aws/provider/ec2	48.829s	coverage: 30.2% of statements
?   	github.com/pulumi/lumi/lib/aws/provider/elasticbeanstalk	[no test files]
?   	github.com/pulumi/lumi/lib/aws/provider/iam	[no test files]
ok  	github.com/pulumi/lumi/lib/aws/provider/lambda	18.071s	coverage: 50.9% of statements
?   	github.com/pulumi/lumi/lib/aws/provider/s3	[no test files]
?   	github.com/pulumi/lumi/lib/aws/provider/testutil	[no test files]
make[1]: *** [test] Error 1
make[1]: Leaving directory `/home/travis/gopath/src/github.com/pulumi/lumi/lib/aws'
make: *** [awspkg] Error 2
```

I think @lukehoban has ideas on how to fix this.  For example, instead of sharing a global name, perhaps each test run should have a unique token of sorts, and then it should only clean up its own mess.  This means possibly accumulating of garbage over time, but it seems better than failures.

--- a/lib/aws/provider/dynamodb/table_test.go
+++ b/lib/aws/provider/dynamodb/table_test.go
@@ -19,27 +19,28 @@ import (
 	"strings"
 	"testing"
 
-	"github.com/stretchr/testify/assert"
-
 	"github.com/aws/aws-sdk-go/aws"
 	awsdynamodb "github.com/aws/aws-sdk-go/service/dynamodb"
 	"github.com/pulumi/lumi/lib/aws/provider/awsctx"
 	"github.com/pulumi/lumi/lib/aws/provider/testutil"
 	"github.com/pulumi/lumi/lib/aws/rpc/dynamodb"
+	"github.com/pulumi/lumi/pkg/resource"
+	"github.com/stretchr/testify/assert"
 )
-
-const RESOURCEPREFIX = "lumitest"
 
 func Test(t *testing.T) {
 	t.Parallel()
 
+	prefix := resource.NewUniqueHex("lumitest", 20, 20)
 	ctx := testutil.CreateContext(t)
-	err := cleanup(ctx)
-	assert.Nil(t, err)
+	defer func() {
+		err := cleanup(prefix, ctx)
+		assert.Nil(t, err)
+	}()
 
 	testutil.ProviderTestSimple(t, NewTableProvider(ctx), TableToken, []interface{}{
 		&dynamodb.Table{
-			Name: aws.String(RESOURCEPREFIX),
+			Name: aws.String(prefix),
 			Attributes: []dynamodb.Attribute{
 				{Name: "Album", Type: "S"},
 				{Name: "Artist", Type: "S"},
@@ -62,7 +63,7 @@ func Test(t *testing.T) {
 			},
 		},
 		&dynamodb.Table{
-			Name: aws.String(RESOURCEPREFIX),
+			Name: aws.String(prefix),
 			Attributes: []dynamodb.Attribute{
 				{Name: "Album", Type: "S"},
 				{Name: "Artist", Type: "S"},
@@ -97,15 +98,15 @@ func Test(t *testing.T) {
 	})
 }
 
-func cleanup(ctx *awsctx.Context) error {
-	fmt.Printf("Cleaning up tables with prefix: %v\n", RESOURCEPREFIX)
+func cleanup(prefix string, ctx *awsctx.Context) error {
+	fmt.Printf("Cleaning up tables with prefix: %v\n", prefix)
 	list, err := ctx.DynamoDB().ListTables(&awsdynamodb.ListTablesInput{})
 	if err != nil {
 		return err
 	}
 	cleaned := 0
 	for _, table := range list.TableNames {
-		if strings.HasPrefix(aws.StringValue(table), RESOURCEPREFIX) {
+		if strings.HasPrefix(aws.StringValue(table), prefix) {
 			if _, delerr := ctx.DynamoDB().DeleteTable(&awsdynamodb.DeleteTableInput{
 				TableName: table,
 			}); delerr != nil {

--- a/lib/aws/provider/ec2/instance_test.go
+++ b/lib/aws/provider/ec2/instance_test.go
@@ -23,9 +23,8 @@ import (
 	"github.com/pulumi/lumi/lib/aws/provider/awsctx"
 	"github.com/pulumi/lumi/lib/aws/provider/testutil"
 	"github.com/pulumi/lumi/lib/aws/rpc/ec2"
+	"github.com/pulumi/lumi/pkg/resource"
 )
-
-const RESOURCEPREFIX = "lumitest"
 
 var amis = map[string]string{
 	"us-east-1":      "ami-6869aa05",
@@ -48,28 +47,28 @@ var amis = map[string]string{
 func Test(t *testing.T) {
 	t.Parallel()
 
+	prefix := resource.NewUniqueHex("lumitest", 20, 20)
 	ctx := testutil.CreateContext(t)
-	cleanup(ctx)
-
+	defer cleanup(prefix, ctx)
 	instanceType := ec2.InstanceType("t2.nano")
 
 	testutil.ProviderTestSimple(t, NewInstanceProvider(ctx), InstanceToken, []interface{}{
 		&ec2.Instance{
-			Name:         aws.String(RESOURCEPREFIX),
+			Name:         aws.String(prefix),
 			InstanceType: &instanceType,
 			ImageID:      amis[ctx.Region()],
 			Tags: &[]ec2.Tag{{
-				Key:   RESOURCEPREFIX,
-				Value: RESOURCEPREFIX,
+				Key:   prefix,
+				Value: prefix,
 			}},
 		},
 		&ec2.Instance{
-			Name:         aws.String(RESOURCEPREFIX),
+			Name:         aws.String(prefix),
 			InstanceType: &instanceType,
 			ImageID:      amis[ctx.Region()],
 			Tags: &[]ec2.Tag{{
-				Key:   RESOURCEPREFIX,
-				Value: RESOURCEPREFIX,
+				Key:   prefix,
+				Value: prefix,
 			}, {
 				Key:   "Hello",
 				Value: "World",
@@ -79,12 +78,12 @@ func Test(t *testing.T) {
 
 }
 
-func cleanup(ctx *awsctx.Context) {
-	fmt.Printf("Cleaning up instances with tag:%v=%v\n", RESOURCEPREFIX, RESOURCEPREFIX)
+func cleanup(prefix string, ctx *awsctx.Context) {
+	fmt.Printf("Cleaning up instances with tag:%v=%v\n", prefix, prefix)
 	list, err := ctx.EC2().DescribeInstances(&awsec2.DescribeInstancesInput{
 		Filters: []*awsec2.Filter{{
-			Name:   aws.String("tag:" + RESOURCEPREFIX),
-			Values: []*string{aws.String(RESOURCEPREFIX)},
+			Name:   aws.String("tag:" + prefix),
+			Values: []*string{aws.String(prefix)},
 		}},
 	})
 	if err != nil {

--- a/lib/aws/provider/lambda/function_test.go
+++ b/lib/aws/provider/lambda/function_test.go
@@ -19,8 +19,6 @@ import (
 	"strings"
 	"testing"
 
-	"github.com/stretchr/testify/assert"
-
 	"github.com/aws/aws-sdk-go/aws"
 	awsiam "github.com/aws/aws-sdk-go/service/iam"
 	awslambda "github.com/aws/aws-sdk-go/service/lambda"
@@ -31,18 +29,20 @@ import (
 	"github.com/pulumi/lumi/lib/aws/rpc/iam"
 	"github.com/pulumi/lumi/lib/aws/rpc/lambda"
 	"github.com/pulumi/lumi/pkg/resource"
+	"github.com/stretchr/testify/assert"
 )
-
-const RESOURCEPREFIX = "lumitest"
 
 func Test(t *testing.T) {
 	t.Parallel()
 
+	prefix := resource.NewUniqueHex("lumitest", 20, 20)
 	ctx := testutil.CreateContext(t)
-	funcerr := cleanupFunctions(ctx)
-	assert.Nil(t, funcerr)
-	roleerr := cleanupRoles(ctx)
-	assert.Nil(t, roleerr)
+	defer func() {
+		funcerr := cleanupFunctions(prefix, ctx)
+		assert.Nil(t, funcerr)
+		roleerr := cleanupRoles(prefix, ctx)
+		assert.Nil(t, roleerr)
+	}()
 
 	sourceARN := rpc.ARN("arn:aws:s3:::elasticbeanstalk-us-east-1-111111111111")
 
@@ -57,7 +57,7 @@ func Test(t *testing.T) {
 				Name: "role",
 				Creator: func(ctx testutil.Context) interface{} {
 					return &iam.Role{
-						Name: aws.String(RESOURCEPREFIX),
+						Name: aws.String(prefix),
 						ManagedPolicyARNs: &[]rpc.ARN{
 							rpc.ARN("arn:aws:iam::aws:policy/AWSLambdaFullAccess"),
 						},
@@ -81,7 +81,7 @@ func Test(t *testing.T) {
 				Name: "f",
 				Creator: func(ctx testutil.Context) interface{} {
 					return &lambda.Function{
-						Name: aws.String(RESOURCEPREFIX),
+						Name: aws.String(prefix),
 						Code: resource.Archive{
 							Assets: &map[string]*resource.Asset{
 								"index.js": {
@@ -99,7 +99,7 @@ func Test(t *testing.T) {
 				Name: "permission",
 				Creator: func(ctx testutil.Context) interface{} {
 					return &lambda.Permission{
-						Name:          aws.String(RESOURCEPREFIX),
+						Name:          aws.String(prefix),
 						Function:      ctx.GetResourceID("f"),
 						Action:        "lambda:InvokeFunction",
 						Principal:     "s3.amazonaws.com",
@@ -115,19 +115,19 @@ func Test(t *testing.T) {
 
 }
 
-func cleanupFunctions(ctx *awsctx.Context) error {
-	fmt.Printf("Cleaning up function with name:%v\n", RESOURCEPREFIX)
+func cleanupFunctions(prefix string, ctx *awsctx.Context) error {
+	fmt.Printf("Cleaning up function with name:%v\n", prefix)
 	list, err := ctx.Lambda().ListFunctions(&awslambda.ListFunctionsInput{})
 	if err != nil {
 		return err
 	}
 	cleaned := 0
 	for _, fnc := range list.Functions {
-		if strings.HasPrefix(aws.StringValue(fnc.FunctionName), RESOURCEPREFIX) {
+		if strings.HasPrefix(aws.StringValue(fnc.FunctionName), prefix) {
 			if _, delerr := ctx.Lambda().DeleteFunction(&awslambda.DeleteFunctionInput{
 				FunctionName: fnc.FunctionName,
 			}); delerr != nil {
-				fmt.Printf("Unable to cleanip function %v: %v\n", fnc.FunctionName, delerr)
+				fmt.Printf("Unable to cleanup function %v: %v\n", fnc.FunctionName, delerr)
 				return delerr
 			}
 			cleaned++
@@ -137,15 +137,15 @@ func cleanupFunctions(ctx *awsctx.Context) error {
 	return nil
 }
 
-func cleanupRoles(ctx *awsctx.Context) error {
-	fmt.Printf("Cleaning up roles with name:%v\n", RESOURCEPREFIX)
+func cleanupRoles(prefix string, ctx *awsctx.Context) error {
+	fmt.Printf("Cleaning up roles with name:%v\n", prefix)
 	list, err := ctx.IAM().ListRoles(&awsiam.ListRolesInput{})
 	if err != nil {
 		return err
 	}
 	cleaned := 0
 	for _, role := range list.Roles {
-		if strings.HasPrefix(aws.StringValue(role.RoleName), RESOURCEPREFIX) {
+		if strings.HasPrefix(aws.StringValue(role.RoleName), prefix) {
 			policies, err := ctx.IAM().ListAttachedRolePolicies(&awsiam.ListAttachedRolePoliciesInput{
 				RoleName: role.RoleName,
 			})

